omegaconf/nodes.py

```diff
@@ -163,6 +164,132 @@ def __deepcopy__(self, memo: Dict[int, Any] = {}) -> "AnyNode":
         return res
 
 
+class UnionNode(ValueNode):
+    def __init__(
+        self,
+        value: Any = None,
+        key: Any = None,
+        parent: Optional[Container] = None,
+        is_optional: bool = True,
+        ref_type: Any = Union[Any],
+        *,
+        element_types: List[Any],
+    ):
+        self.element_types = element_types
+        super().__init__(
+            parent=parent,
+            value=value,
+            metadata=Metadata(
+                key=key,
+                optional=is_optional,
+                ref_type=ref_type,
+                object_type=type(value),
+            ),
+        )
+
+    def _wrap_node(
+        self,
+        value: Any,
+        key: Any = None,
+        is_optional: bool = True,
+        parent: Optional[BaseContainer] = None,
+        ref_type: Any = None,
+        use_type: bool = False,
+    ) -> Optional[Node]:
+        from omegaconf.omegaconf import _maybe_wrap
+
+        if value is None:
+            return None
+        else:
+            return _maybe_wrap(
+                ref_type=ref_type,
+                key=key,
+                value=value,
+                is_optional=is_optional,
+                parent=parent,
+                use_type=use_type,
+            )
+
+    def validate_and_convert(self, value: Any) -> Optional[Any]:
+        from omegaconf._utils import is_structured_config
+        from omegaconf.basecontainer import BaseContainer
+
+        value_node = None
+        valid_value = False
+        value_type: Any = type(value)
+        if isinstance(value, BaseContainer):
+            ref_type = value._metadata.ref_type
+            if is_structured_config(ref_type):
+                value_type = ref_type
+            if any(
+                issubclass(value_type, union_type) for union_type in self.element_types
```

Comment:
I've been adding tests and it fails as you pointed out. The code you wrote could be useful, I'll try it. 
I'm finishing a sprint this week so sometime this week i expect to have a fix.

Reply:
So I've been adding tests and encountered wrapping like this: 
```python
node = DictConfig({"foo": "invalid"})
_maybe_wrap(value=node._value(), ref_type=Dict[str, int], is_optional=True, key=None, parent=    None)
```
which fails because "invalid" is AnyNode and validate_set doesn't take it well. I solved this in #443 so this might need to wait that PR.

Reply:
@odelalleau, ref_type is the annotated type.
for example:
```
@dataclass
class Foo:
 x : BaseClass = SubClass()
```

BaseClass is the ref_type.
Subclass is the type.
ref_type is checked against during assignment, for example.

Reply:
> So I've been adding tests and encountered wrapping like this:
> 
> ```python
> node = DictConfig({"foo": "invalid"})
> _maybe_wrap(value=node._value(), ref_type=Dict[str, int], is_optional=True, key=None, parent=    None)
> ```
> 
> which fails because "invalid" is AnyNode and validate_set doesn't take it well. I solved this in #443 so this might need to wait that PR.

Not sure I understand, isn't it expected to fail? (since `"invalid"` is not an `int`)

Reply:
It is expected to fail. and it is fixed in that other PR @pereman2  mentioned.